### PR TITLE
Add fn return type and some more validations

### PR DESCRIPTION
**Pull request by barakb**
## Summary by CodeRabbit

* **New Features**
  * Granular extraction of temporal/duration components via property access.
  * Per-function return-type metadata to improve predicate/type resolution.
  * Path-variable property access allowed in SET targets; SET on Null/non-entity types is ignored.
  * Pre-registration of path variables during graph binding for more consistent resolution.

* **Bug Fixes**
  * Runtime validation enforcing boolean predicates across patterns, calls, projections and list comprehensions.
  * COUNT(DISTINCT *) now triggers an early error; clearer inlined-property error reporting.

* **Tests**
  * Re-enabled and expanded query validation tests covering many edge cases.

* **Chores**
  * Updated chrono dependency version.

### CODE REVIEW

File: graph/src/runtime/runtime.rs [hunk 1]
```diff
@@ -2825,6 +2837,223 @@ impl<'a> Runtime {
...
+    /// epoch).  Mirrors the C `DateTime_getComponent` in `datetime.c`.
+    fn get_datetime_component(
+        timestamp_ms: i64,
+        component: &str,
+    ) -> Result<Value, String> {
+        use chrono::{Datelike, TimeZone, Timelike, Utc};
+
+        let chrono::LocalResult::Single(dt) = Utc.timestamp_millis_opt(timestamp_ms) else {
+            return Ok(Value::Null);
+        };
```
[the remark this answers is not in the document]
Comment (barakb) [reply graph/src/runtime/runtime.rs]: fixed

**Comment by barakb** [comment]
@coderabbitai review

File: graph/src/runtime/runtime.rs [hunk 2]
```diff
@@ -2825,6 +2833,228 @@ impl<'a> Runtime {
...
+            timestamp_ms % 1000
+        } else if c.eq_ignore_ascii_case("microsecond") || c.eq_ignore_ascii_case("nanosecond") {
+            // microsecond/nanosecond precision not stored
+            0
+        } else {
+            return Err(format!("unknown datetime component {component}"));
+        };
+
+        Ok(Value::Int(val))
+    }
```
[the remark this answers is not in the document]
Comment (barakb) [reply graph/src/runtime/runtime.rs]: fixed

File: graph/src/binder.rs [hunk 3]
```diff
@@ -124,6 +124,9 @@ impl Binder {
             } => {
                 let pattern = self.bind_graph(pattern, false)?;
                 let filter = filter.map(|expr| self.bind_expr(&expr)).transpose()?;
+                if let Some(ref f) = filter {
```
Comment (DvirDukhan) [inline graph/src/binder.rs]: or else? what happens if filter is None?

File: graph/src/binder.rs [hunk 4]
```diff
@@ -588,6 +607,22 @@ impl Binder {
...
+    fn bind_set_target(
+        &mut self,
+        expr: &QueryExpr<Arc<String>>,
+    ) -> Result<QueryExpr<Variable>, String> {
+        let root = expr.root();
+        Ok(Arc::new(self.bind_expr_node_impl(
+            expr,
+            root,
+            &mut vec![],
+            true,
```
Comment (DvirDukhan) [inline graph/src/binder.rs]: we should return an error in case of path property

File: graph/src/binder.rs [hunk 5]
```diff
@@ -849,4 +883,99 @@ impl Binder {
...
+            | ExprIR::Le
+            | ExprIR::Ge
+            | ExprIR::In
+            | ExprIR::IsNode
+            | ExprIR::IsRelationship
+            | ExprIR::Quantifier(_, _)
+            | ExprIR::Variable(_)
+            | ExprIR::Parameter(_)
+            | ExprIR::Property(_) => true,
+        }
```
[the remark this answers is not in the document]
Comment (barakb) [reply graph/src/binder.rs]: WHERE NULL is false in cypher, Variable, Parameter, Property, Quantifier may return boolean this function is compile time type checking that with our type checking system can not at compile time find the type of every expression, so it only reject expression that can prove at compile time not to have a boolean type.
I will change the name of the function to expr_may_return_boolean to be more direct with that

File: graph/src/binder.rs (hunk 4 above)
Comment (barakb) [reply graph/src/binder.rs]: fixed

File: graph/src/binder.rs [hunk 6]
```diff
@@ -849,4 +912,99 @@ impl Binder {
         }
         Ok(())
     }
+
+    /// Validates that a filter expression can return a boolean value.
+    /// This mirrors `_FilterTree_ValidExpressionNode` / `AR_EXP_ReturnsBoolean`
+    /// from the C implementation and must be called at bind time for all
+    /// WHERE predicates (Match, With, Call) and list-comprehension WHERE clauses.
+    fn validate_filter_predicate(expr: &QueryExpr<Variable>) -> Result<(), String> {
```
[the remark this answers is not in the document]
Comment (barakb) [reply graph/src/binder.rs]: Done

**Review by AviAvni — APPROVED** [review]
